File: 78-subsets/78-subsets.scala
import scala.collection.mutable._

object Solution {
  def subsets(nums: Array[Int]): List[List[Int]] = {
    val res: ListBuffer[List[Int]] = ListBuffer()
    helper(nums, 0, ListBuffer(), res)
    res.toList
  }

  def helper(
      nums: Array[Int],
      idx: Int,
      current: ListBuffer[Int],
      res: ListBuffer[List[Int]]
  ): Unit = {
    if (idx >= nums.length) {
      res += current.toList
    } else {
      // not take
      helper(nums, idx + 1, current, res)

      // take
      current += nums(idx)
      helper(nums, idx + 1, current, res)
      current.remove(current.length - 1)
    }
  }
}

File: 78-subsets/78-subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        helper(nums, 0, result, new LinkedList<>());
        return result;
    }
    
    void helper(int[] nums, int n, List<List<Integer>> result, Deque<Integer> stack) {
        if (n == nums.length) {
            result.add(List.copyOf(stack));
            return;
        }
        
        stack.push(nums[n]);
        helper(nums, n + 1, result, stack);
        stack.pop();
        helper(nums, n + 1, result, stack);
    }
}